File: leetcode/221/c++/solution_1.cpp
/*
    221. Maximal Square
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    int maximalSquare(vector<vector<char>>& matrix) {
        vector<vector<int>> num_matrix(matrix.size(),
                                       vector<int>(matrix[0].size(), 0));
        int solution = 0, min_local_val = 0;
        int x[4] = {-1, 0, -1};
        int y[4] = {-1, -1, 0};
        for (int i = 0; i < matrix.size(); ++i) {
            for (int j = 0; j < matrix[i].size(); ++j) {
                num_matrix[i][j] = matrix[i][j] - '0';
                if (num_matrix[i][j] != 0) {
                    min_local_val = -1;
                    for (size_t k = 0; k < 3; ++k) {
                        if (i + x[k] >= 0 && j + y[k] >= 0 &&
                            num_matrix[i + x[k]][j + y[k]] != 0) {
                            min_local_val =
                                min_local_val == -1
                                    ? num_matrix[i + x[k]][j + y[k]]
                                    : (num_matrix[i + x[k]][j + y[k]] <
                                               min_local_val
                                           ? num_matrix[i + x[k]][j + y[k]]
                                           : min_local_val);
                        } else {
                            min_local_val = 0;
                            break;
                        }
                    }
                    num_matrix[i][j] = min_local_val + 1;
                    if (solution < num_matrix[i][j]) {
                        solution = num_matrix[i][j];
                    }
                }
            }
        }

        return solution * solution;
    }
};

File: leetcode/1930/rust/solution_1.rs
/*
    1930. Unique Length-3 Palindromic Subsequences
    Author: Miguel Angel Bermeo Ayerbe
*/
impl Solution {
    pub fn count_palindromic_subsequence(s: String) -> i32 {
        let mut middle_marks: [i32; 26] = [0; 26];
        let mut last: [usize; 26] = [0; 26];
        let mut first: [usize; 26] = [0; 26];
        for (i, char_i) in s.chars().enumerate() {
            let char_b_i: usize = ((char_i as u8) - b'a') as usize;
            if 2 <= i {
                if first[char_b_i] != 0 {
                    for i in 0..26 {
                        if last[i] != 0 && first[char_b_i] < last[i] {
                            middle_marks[char_b_i] |= 1 << i;
                        }
                    }
                }
            }
            last[char_b_i] = i + 1;
            if first[char_b_i] == 0 {
                first[char_b_i] = i + 1;
            }
        }
        let mut output: i32 = 0;
        for i in 0..26 {
            if first[i] > 0 {
                for j in 0..26 {
                    output += (middle_marks[i] >> j) & 1;
                }
            }
        }
        output
    }
}
